src/flashcart/ed64/ed64.c

```diff
@@ -0,0 +1,288 @@
+    //     }
+    //     if (ed64_flash_wait_busy() != ED64_OK) {
+    //         return FLASHCART_ERROR_INT;
+    //     }
+    //     address += program_size;
+    //     size -= program_size;
+    //     *br += bp;
+    // }
+
+//     return FLASHCART_OK;
+// }
+
+static void load_cleanup (FIL *fil) {
+    ed64_bios_byteswap_off();
+    f_close(fil);
+}
+
+
+static flashcart_error_t ed64_init (void) {
+    // uint16_t major;
+    // uint16_t minor;
+
+    // HACK: Because libcart reads PI config from address 0x10000000 when initializing
+    //       we need to write safe value before running any libcart function.
+    //       Data in SDRAM can be undefined at this point and result in setting incorrect PI config.
+    extern uint32_t cart_dom1;
+    cart_dom1 = 0x80371240;
+
+    // ed64_bios_unlock_regs();
+
+    ed64_bios_init();
+    // ed64_bios_reset_spx(); // only required if V2+ FW is updating!
+
+    ed64_bios_set_ram_bank(1); // Seemingly bank 0 (for SRAM 128k) is only supported on V3 and it is not the default.
+    ed64_bios_gpio_mode_io();
+
+    // if (!ed64_check_presence()) {
+    //     return FLASHCART_ERROR_UNSUPPORTED;
+    // }
+    // if (ed64_get_version(&major, &minor) != ED64_OK) {
+    //     return FLASHCART_ERROR_OUTDATED;
+    // }
+    // if (major != SUPPORTED_MAJOR_VERSION) {
+    //     return FLASHCART_ERROR_OUTDATED;
+    // }
+    // if (minor < SUPPORTED_MINOR_VERSION) {
+    //     return FLASHCART_ERROR_OUTDATED;
+    // }
+
+    // bool writeback_pending;
+    // do {
+    //     if (ed64_writeback_pending(&writeback_pending) != ED64_OK) {
+    //         return FLASHCART_ERROR_INT;
+    //     }
+    // } while (writeback_pending);
+
+    // const struct {
+    //     ed64_cfg_t id;
+    //     uint32_t value;
+    // } default_config[] = {
+    //     { CFG_BOOTLOADER_SWITCH, false },
+    //     { CFG_ROM_WRITE_ENABLE, false },
+    //     { CFG_ROM_SHADOW_ENABLE, false },
+    //     { CFG_DD_MODE, DD_MODE_DISABLED },
+    //     { CFG_ISV_ADDRESS, 0x00000000 },
+    //     { CFG_BOOT_MODE, BOOT_MODE_MENU },
+    //     { CFG_SAVE_TYPE, SAVE_TYPE_NONE },
+    //     { CFG_CIC_SEED, CIC_SEED_AUTO },
+    //     { CFG_TV_TYPE, TV_TYPE_PASSTHROUGH },
+    //     { CFG_DD_SD_ENABLE, false },
+    //     { CFG_DD_DRIVE_TYPE, DRIVE_TYPE_RETAIL },
+    //     { CFG_DD_DISK_STATE, DISK_STATE_EJECTED },
+    //     { CFG_BUTTON_MODE, BUTTON_MODE_NONE },
+    //     { CFG_ROM_EXTENDED_ENABLE, false },
+    // };
+
+    // for (int i = 0; i < sizeof(default_config) / sizeof(default_config[0]); i++) {
+    //     if (ed64_set_config(default_config[i].id, default_config[i].value) != ED64_OK) {
+    //         return FLASHCART_ERROR_INT;
+    //     }
+    // }
+
+    return FLASHCART_OK;
+}
+
+static flashcart_error_t ed64_deinit (void) {
+    // // NOTE: Necessary because libcart enables ROM write by default
+    // ed64_set_config(CFG_ROM_WRITE_ENABLE, false);
+
+    ed64_bios_lock_regs();
```

Comment:
Shouldn't locking registers be the last step of deinitialization?

Reply:
This needs a bit of further thought (at least for the "gpio_mode". Depending on whether the flashcart in use has a RTC. For V3 only, the mode should be set to gpio_mode_rtc() and for others gpio_mode_off(). In the official menu RTC is only enabled if the ROM requires it (Doubutsu no Mori) and others that are explicitly set. 

Anyway, the deinit function is the wrong place for it to be configured.